src/operators/join/hash_join.cc

```diff
@@ -0,0 +1,325 @@
+                   std::shared_ptr<OperatorOptions> options)
+    : Operator(query_id, options),
+      prev_result_(prev_result),
+      output_result_(output_result),
+      predicate_(predicate) {
+    // prev result has two operator result - 0 - left, 1 - right
+  assert(prev_result.size() == 2);
+  joined_indices_.resize(2);
+}
+
+void HashJoin::execute(Task *ctx) {
+  ctx->spawnTask(CreateTaskChain(
+      CreateLambdaTask([this](Task *internal) { Initialize(internal); }),
+      CreateLambdaTask([&, this](Task *internal) { Join(internal); }),
+      CreateLambdaTask([&, this](Task *internal) { Finish(); })));
+}
+
+void HashJoin::Initialize(Task *ctx) {
+  left_table_ =
+      prev_result_.at(LEFT_JOIN_REF_IDX)->get_table(predicate_->left_col_.table);
+  right_table_ =
+      prev_result_.at(RIGHT_JOIN_REF_IDX)->get_table(predicate_->right_col_.table);
+
+  left_table_.MaterializeColumn(ctx, predicate_->left_col_.col_name, lcol_);
+  right_table_.MaterializeColumn(ctx, predicate_->right_col_.col_name, rcol_);
+}
+
+void HashJoin::Join(Task *ctx) {
+  ctx->spawnTask(CreateTaskChain(
+      CreateLambdaTask([&, this](Task *internal) {
+        // Build phase
+        if (right_table_.hash_table() != nullptr) {
+          hash_table_ = right_table_.hash_table();
+        } else {
+          BuildHashTable(
+              rcol_.chunked_array(),
+              (right_table_.filter.kind() == arrow::Datum::CHUNKED_ARRAY)
+                  ? right_table_.filter.chunked_array()
+                  : nullptr,
+              internal);
+        }
+      }),
+      CreateLambdaTask([&, this](Task *internal) {
+        ProbeHashTable(lcol_.chunked_array(), left_table_.filter,
+                       left_table_.indices, internal);
+      }),
+      CreateLambdaTask([this](Task *internal) { FinishProbe(internal); })));
+}
+
+void HashJoin::BuildHashTable(
+    const std::shared_ptr<arrow::ChunkedArray> &col,
+    const std::shared_ptr<arrow::ChunkedArray> &filter, Task *ctx) {
+  hash_table_ =
+      std::make_shared<phmap::flat_hash_map<int64_t, std::vector<RecordID>>>();
+
+  // Precompute the row offsets of each chunk. A multithreaded build phase
+  // requires that we know all offsets beforehand.
+  std::vector<uint64_t> chunk_offsets(col->num_chunks());
+  ComputeChunkOffsets(chunk_offsets, col);
+
+  size_t hash_table_size = col->length();
+  if (filter != nullptr) {
+    uint32_t updated_length = 0;
+
+    for (int j = 0; j < col->num_chunks(); ++j) {
+      updated_length += arrow::compute::internal::GetFilterOutputSize(
+          *filter->chunk(j)->data(),
+          arrow::compute::FilterOptions::NullSelectionBehavior::DROP);
+    }
+    hash_table_size = updated_length;
+  }
+  hash_table_->reserve(hash_table_size);
+
+  for (std::size_t i = 0; i < col->num_chunks(); i++) {
+    // Each task inserts one chunk into the hash table
+    // TODO: for now, we assume the join column is INT64 type.
```

Comment:
Got it - will attemp to generalize this in later PR.

Reply:
No issues, I am planning to work on providing support multiple types in Join in the coming PRs.